remme/tp/node_account.py

```diff
@@ -222,35 +226,60 @@ def _transfer_from_unfrozen_to_operational(self, context, node_account_public_ke
-        node_account_address = self.make_address_from_data(node_account_public_key)
+    def _transfer_from_frozen_to_unfrozen(self, context, public_key, payload):
+        node_account_address = self.make_address_from_data(public_key)
 
         node_account = get_data(context, NodeAccount, node_account_address)
-        minimum_stake = _get_setting_value(context, 'remme.settings.minimum_stake')
+        if node_account.node_state != NodeAccount.OPENED:
+            raise InvalidTransaction('Masternode is not opened or has been closed.')
 
-        if node_account is None:
-            raise InvalidTransaction('Invalid context or address.')
+        min_stake = _get_setting_value(context, SETTINGS_MINIMUM_STAKE)
+        if min_stake is None or not min_stake.isdigit():
+            raise InvalidTransaction(f'{SETTINGS_MINIMUM_STAKE} is malformed. Not set.')
 
-        if minimum_stake is None or not minimum_stake.isdigit():
-            raise InvalidTransaction('Wrong minimum stake address.')
+        min_stake = client_to_real_amount(int(min_stake))
 
-        minimum_stake = client_to_real_amount(int(minimum_stake))
+        if node_account.reputation.frozen <= min_stake:
+            raise InvalidTransaction("Not enough tokens on frozen balance to defrost")
 
-        amount = client_to_real_amount(internal_transfer_payload.value)
+        ledger_tax = _get_setting_value(context, SETTINGS_BLOCKCHAIN_TAX)
+        if ledger_tax is None:
+            raise InvalidTransaction(f'{SETTINGS_BLOCKCHAIN_TAX} is malformed. Not set.')
+
+        max_share = Decimal(1 - Decimal(ledger_tax))
+
+        now = datetime.utcnow().timestamp()
+        tokens = 0
+        for share in reversed(node_account.shares):
+            if share.defrost_months == 12:
+                continue
 
-        if node_account.reputation.frozen < minimum_stake:
-            raise InvalidTransaction('Frozen balance is lower than the minimum stake.')
+            delta = datetime.fromtimestamp(now) - datetime.fromtimestamp(share.block_timestamp)
 
-        if node_account.reputation.frozen - amount < minimum_stake:
-            raise InvalidTransaction('Frozen balance after transfer lower than the minimum stake.')
+            # determine what was the diferrence
+            share_that_should_defrost = min(delta.days // 28, 12)
+            if not share_that_should_defrost or share.defrost_months >= share_that_should_defrost:
+                continue
 
-        node_account.reputation.frozen -= amount
-        node_account.reputation.unfrozen += amount
+            demand_month = (share_that_should_defrost - share.defrost_months)
+
+            defrosted_share = self._calculate_share_to_defrost(real_to_client_amount(share.frozen_share), demand_month, max_share)
+            defrosted_tokens = client_to_real_amount(defrosted_share * share.reward, 0)
+            tokens += defrosted_tokens
+
+            share.defrost_months += share_that_should_defrost
```

Comment:
Example:
share.defrost_month = 3
share_that_should_defrost = 4
demand_month = 1 (share_that_should_defrost - share.defrost_month)

but after found count of tokens for demand_month (1) you do next:
share.defrost_months += share_that_should_defrost = 7 , but should be 4, because we defrost for demand_month that equal to 1

So should be some of this:
share.defrost_month = share_that_should_defrost
or
share.defrost_month += demand_month